xymon-test/tools/tr2latex/tr2latex.man
.TH "TR2TEX" "1" "25 August 2008" "" ""
.SH "NAME"
tr2latex \- convert a document from troff to LaTeX
.SH "SYNOPSIS"
tr2latex
[\fB\-m\fR]
[\fB\-t\fR]
[\fB\-n\fR]
[\fB\-b\fR]
[\fB\-s style\fR]
[\fB\-o outfile\fR]
filename ...
.SH "DESCRIPTION"
.B Tr2latex
converts a document typeset in
.B troff
to a
.B LaTeX
format.
It is intended to do the first pass of the conversion. The user
should then finish up the rest of the conversion and customize the
converted manuscript to his/her liking.
It can also serve as a tutor for those who want to convert from
troff to LaTeX.
.PP 
Most of the converted document will be in LaTeX
but some of it may
be in plain
.B TeX.
It will also use some macros in
.B troffms.sty
or
.B troffman.sty
which are included in the package and must be available to the document
when processed with LaTeX.
.PP 
If there is more than one input file, they will all be converted into
one LaTeX document.
.PP 
.B Tr2latex
understands most of the
.B \-ms
and
.B \-man
macros and
.B eqn
preprocessor symbols. It also understands several plain
.B troff
commands. Few
.B tbl
preprocessor commands are understood to help convert very simple tables.
.PP 
When converting manuals, use the
.B \-m
flag. Otherwise
.B tr2latex
assumes to translate a
.B \-me
input file.
.PP 
If a troff command cannot be converted, the line that contain that
command will be commented out.
.PP 
NOTE: if you have
.B eqn
symbols, you must have the in\-line mathematics delimiter defined by
.B delim
in the file you are converting. If it is defined in another
setup file, that setup file has to be concatenated with the
file to be converted, otherwise
.B tr2latex
will regard the in\-line math as ordinary text.
.SH "OPTIONS"
.TP 12
.B \-m
Convert manual pages. This makes
.B tr2latex
understand most of the
.B \-man
macros. It uses the style file
.BI troffman.sty .
.TP 
.B \-t
Produce twoside page style.
.TP 
.B \-\fIn\fP
Use a font size of
.I R n pt.
The default font size is 12pt for
.B \-man
and 11pt otherwise.
.TP 
.B "\-s \fIstyle\fP"
Use the style file
.I style
instead of the default
.BR article.sty .
.TP 
.B "\-o \fIoutfile\fP"
Write output to file
.I R outfile .
.TP 
\fB\-b\fR
print only the LaTeX body. Usefull when converting manpages which are included in a larger document.
.SH "BUGS"
Many of these bugs are harmless. Most of them cause local errors
that can be fixed in the converted manuscript.
.PP 
\- Some macros and macro arguments are not recognized.
.PP 
\- Commands that are not separated from their argument by a space are
not properly parsed (e.g .sp3i).
.PP 
\- When some operators (notably over, sub and sup) are renamed (via define),
then they are encountered in the text,
.B tr2latex
will treat them as
ordinary macros and will not apply their rules.
.PP 
\- rpile, lpile and cpile are treated the same as cpile.
.PP 
\- rcol, lcol are treated the same as ccol.
.PP 
\- Math\-mode size, gsize, fat, and gfont are ignored.
.PP 
\- lineup and mark are ignored. The rules are so different.
.PP 
\- Some troff commands are translated to commands that require
delimiters that have to be explicitly put. Since they are
sometimes not put in troff, they can create problems.
Example: .nf not closed by .fi .
.PP 
\- When local motions are converted to \\raise or \\lower, an \\hbox
is needed, which has to be put manually after the conversion.
.PP 
\- 'a sub i sub j' is converted to 'a_i_j' which TeX
parses as 'a_i{}_j}' with a complaint that it is vague. 'a sub {i sub j}'
is parsed correctly and converted to 'a_{i_j}'.
.PP 
\- Line spacing is not changed within a paragraph in TeX
(which is a bad practice anyway).
TeX uses the last line spacing in effect in that paragraph.
.SH "TODO"
.TP 
Access registers via
\fB.nr\fR
command.
.TP 
handling of tables
.TP 
handling of tabs
.TP 
\fB.bd\fR
.TP 
\fB.pl\fR
.TP 
\fB.ps\fR
.TP 
\fB.ds\fR
.SH "FILES"
$(TEXLIB)/macros/troffman.sty
.br 
$(TEXLIB)/macros/troffms.sty
.SH "AUTHOR"
Kamal Al\-Yahya, Stanford University
.br 
Christian Engel, Aachen University of Technology


Changes
Changes from 4.3.0 beta 1 -> 4.3.0 beta 2 (24 Apr 2009)
=======================================================
* New "--shuffle" option for bbtest-net to run network tests
  in a random order.
* Client startup script now exports important environment variables,
  so they are actually used in systems with a traditional shell.
* hobbitlaunch no longer crashes if there are no tasks
* Client configure script includes librt for clock_gettime()
* New client support for mainframes: z/OS, z/VM, z/VSE
* Enhanced eventlog and top-changing hosts webpages.
* Revert debian package pathnames back to use "hobbit", so
  updates from 4.2.x will actually work.
* Ghostlist options in hobbitcgi.cfg had no effect because
  of typo in setting name.
* "data" messages could crash hobbitd.
* Debug output from hobbitd_channel now logs only the relevant
  data instead of the full message.
* trimhistory now informs the history module to re-open the
  "allevents" file after trimming it.
* devmon template fix
* New "rrdcachectl" utility included.
* Fixed sorting routine (affected holiday list and others)
* Fix generic crash in communications module between Xymon
  programs, where an empty response message would crash caller.

Changes from 4.2.3 -> 4.3.0 beta 1 (09 Feb 2009)
================================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------

xymon-test/tools/tr2latex/tr2latex.c
/*
** tr2latex - troff to LaTeX converter
** COPYRIGHT (C) 1987 Kamal Al-Yahya, 1991,1992 Christian Engel
**
** Module: tr2latex.c
**
** This module contains the main function inititating the translation
** and supporting the Usage.
*/

#include <getopt.h>

#define MAIN
#include	"setups.h"
#include	"protos.h"

const char *version = "2.3";

int	man,		/* option -m: manual page */
  fontsize,	/* option -9/-10/-11/-12: font size */
  twoside,	/* option -t: twoside */
  piped_in;

char *document = "article";	/* document type, see also -s option */

FILE *out_file;		/* in case they can't use UN*X redirecting or piping */

char *prgname;
char inbuf[MAXLEN],
  outbuf[MAXLEN];

static void usage (int exitcode)
{
  printf ("tr2latex (c) 1986/1987 Kamal Al-Yahya, 1991 C. Engel, 2008..2009 Dirk Jagdmann\n"
	  "Version %s\n",
	  version);
  printf ("tr2latex - troff to LaTeX converter\n"
	  "SYNTAX:  tr2latex [-m] [-t] [-<n>] [-s <style>] [-o <outfile>] [-d] [-b] [<file>...]\n"
	  "options: -m            for manual\n"
	  "         -t            twoside page style\n"
	  "         -<n>          a number n gives the font size (default is 12pt\n"
	  "                       for man, 11pt otherwise)\n"
	  "         -s <style>    use documentstyle <style> (default is article)\n"
	  "         -o <outfile>  send output to <outfile> (default is stdout)\n"
	  "         -d            debug output\n"
	  "         -b            LaTeX body, don't print declarations\n"
	  );

  exit (exitcode);
}

static void process (FILE *in_file, char *f_name, char *pin, char *pout)
{
  static char sep[] = "--------------------------------------------------";

  tmpbuf (in_file, pin);
  fprintf (out_file, "%%%s\n%% start of input file: %s\n%%\n", sep, f_name);
  troff_tex (pin, pout, 0, 0);
  fputs (pout, out_file);
  fprintf (out_file, "%%\n%% end of input file: %s\n%%%s\n", f_name, sep);
}

int main (int argc, char *argv[])
{
  char *pin = inbuf,
    *pout = outbuf;
  FILE *in_file;
  long timeval;		/* clock value from time() for ctime()	*/
  int c, only_body=0;

  prgname = argv [0];
  out_file = stdout;		/* default output */

  /* process option flags */
#if 0
  getopts (&argc, argv);
#else

  while((c=getopt(argc, argv, "0123456789bhmtds:o:")) != EOF)
    switch(c)
      {
      default:
      case '?':
      case 'h': usage (EXIT_SUCCESS);

      case 'b': only_body=1; break;
      case 'm': man = 1; break;
      case 't': twoside = 1; break;
      case 's': document = optarg; break;
      case 'o':
	if ((out_file = fopen(optarg, "w")) == NULL)
	  {
	    fprintf(stderr, "%s: can't open %s\n", prgname, optarg);
	    usage (EXIT_FAILURE);
	  }
	break;
      case 'd': debug_o = 1; break;

      case '0': fontsize = 10; break;
      case '1': fontsize = 11; break;
      case '2': fontsize = 12; break;
      case '3': fontsize = 13; break;
      case '4': fontsize = 14; break;
      case '5': fontsize = 15; break;
      case '6': fontsize = 16; break;
      case '7': fontsize = 17; break;
      case '8': fontsize = 18; break;
      case '9': fontsize = 9; break;
      }
#endif

  /* initialize spacing and indentation parameters */
  strcpy(linespacing.def_units,"\\normalbaselineskip");
  strcpy(linespacing.old_units,"\\normalbaselineskip");
  strcpy(indent.def_units,"em");
  strcpy(indent.old_units,"em");
  strcpy(tmpind.def_units,"em");
  strcpy(tmpind.old_units,"em");
  strcpy(space.def_units,"\\baselineskip");
  strcpy(space.old_units,"\\baselineskip");
  strcpy(vspace.def_units,"pt");
  strcpy(vspace.old_units,"pt");
  linespacing.value = 1.;
  linespacing.old_value = 1.;
  indent.value = 0.;
  indent.old_value = 0.;
  tmpind.value = 0.;
  tmpind.old_value = 0.;
  space.value = 1.;
  space.old_value = 1.;
  vspace.value = 1.;
  vspace.old_value = 1.;
  linespacing.def_value = 0;
  indent.def_value = 0;
  tmpind.def_value = 0;
  space.def_value = 1;
  vspace.def_value = 1;

  math_mode = 0;					/* start with non-math mode */
  de_arg = 0;                     /* not a .de argument */

  /* start of translated document */

  timeval = time(0);
  fprintf (out_file,
	   "%% -*-LaTeX-*-\n\
%% Converted automatically from troff to LaTeX\n\
%% by tr2latex %s\n\
%% on %s\
%% tr2latex was written by Kamal Al-Yahya at Stanford University <Kamal%%Hanauma@SU-SCORE.ARPA>\n\
%% and substantially enhanced by Christian Engel at RWTH Aachen <krischan@informatik.rwth-aachen.de>\n\
%%\n\
%% troff input file%s:%s",
	   version, ctime(&timeval),
	   argc>2?"s":"",
	   argc==1?" <stdin>":"");
#if 0
  for (argi = 1; argi < argc; argi++)
    {
      if (strcmp (argv [argi], "-") == 0)
	fprintf (out_file, " <stdin>");
      else
	fprintf (out_file, " %s", argv[argi]);
    }
#endif

  if(!only_body)
    {
      /* document style and options */
      fprintf (out_file,"\n\n\\documentclass[");
      if (fontsize == 0 && !man)
	fontsize = 11;
      if (fontsize != 0)
	fprintf (out_file,"%dpt", fontsize);
      if (twoside)
	fputs (",twoside", out_file);
      fprintf (out_file,"]{%s}", document);
      fprintf (out_file,"\n\n\\usepackage{%s}", man? "troffman": "troffms");
      fprintf (out_file,"\n\n\\begin{document}\n");
    }

  if (optind == argc)
    process (stdin, "<stdin>", pin, pout);
  else
    {
      for (; optind!=argc; ++optind)
	{
	  if (strcmp (argv[optind], "-") == 0)
	    process (stdin, "<stdin>", pin, pout);
	  else if ((in_file = fopen(argv[optind],"r")) == NULL)
	    fprintf(stderr,"%s: Cannot open input file `%s'\n",
		    prgname,argv[optind]);
	  else
	    {
	      process (in_file, argv[optind], pin, pout);
	      fclose(in_file);
	    }
	}
    }

  if(!only_body)
    {
      /* close translated document */
      fputs("\\end{document}\n",out_file);
    }

  exit(EXIT_SUCCESS);
}

void errexit (int exitcode)
{
  fprintf (stderr, "%s: Error #%03d ", prgname, exitcode);
  exit (exitcode);
}
